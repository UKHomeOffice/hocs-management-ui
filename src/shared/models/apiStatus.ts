export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
