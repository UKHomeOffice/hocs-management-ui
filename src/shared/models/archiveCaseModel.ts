export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
