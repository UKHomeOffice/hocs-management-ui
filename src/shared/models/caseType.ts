export default interface CaseType {
    displayName: string;
    label: string;
    shortCode: string;
    type: string;
    value: string;
}
