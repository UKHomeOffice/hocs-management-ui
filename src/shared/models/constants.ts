export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
