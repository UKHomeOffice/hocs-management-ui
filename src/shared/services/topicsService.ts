import axios from 'axios';
import Item from '../models/item';

export const addChildTopic = (parentTopicId: string, displayName: string) => new Promise((resolve, reject) => axios
    .post(`/api/topics/parents/${parentTopicId}`, { displayName })
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const getParentTopics = () => new Promise((resolve, reject) => axios
    .get<Item[]>('/api/topics/parents')
    .then(response => resolve(response.data.sort((a, b) => {
        const labelA = a.label.toLocaleLowerCase();
        const labelB = b.label.toLocaleLowerCase();
        let comparison = 0;
        if (labelA > labelB) {
            comparison = 1;
        } else if (labelA < labelB) {
            comparison = -1;
        }
        return comparison;
    })))
    .catch(reason => reject(reason))
);
