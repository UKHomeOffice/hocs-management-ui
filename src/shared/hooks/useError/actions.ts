import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
