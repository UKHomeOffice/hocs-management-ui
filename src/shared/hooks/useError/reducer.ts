import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
