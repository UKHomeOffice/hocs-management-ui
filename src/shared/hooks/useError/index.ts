import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
