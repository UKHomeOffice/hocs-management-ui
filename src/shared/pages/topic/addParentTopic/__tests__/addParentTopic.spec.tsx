import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import * as TopicsService from '../../../../services/topicsService';
import { GENERAL_ERROR_TITLE, VALIDATION_ERROR_TITLE, DUPLICATE_PARENT_TOPIC_DESCRIPTION, ADD_PARENT_TOPIC_ERROR_DESCRIPTION } from '../../../../models/constants';
import * as useError from '../../../../hooks/useError';
import AddParentTopic from '../addParentTopic';
import ErrorMessage from '../../../../models/errorMessage';

let match: match<any>;
let history: History<any>;
let location: Location;
let wrapper: RenderResult;

const useStateSpy = jest.spyOn(React, 'useState');
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();
const setDisplayNameSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <AddParentTopic history={history} location={location} match={match}></AddParentTopic>
    </MemoryRouter>
);

jest.mock('../../../../services/topicsService', () => ({
    __esModule: true,
    addParentTopic: () => Promise.resolve()
}));

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: {},
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };

    useStateSpy.mockImplementation(() => ['displayName', setDisplayNameSpy]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the addParentTopic component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the topic name is entered', () => {
    it('should be persisted in the page state', async () => {
        expect.assertions(1);

        const displayNameElement = await waitForElement(async () => {
            return await wrapper.findByLabelText('Topic Name');
        });

        fireEvent.change(displayNameElement, { target: { name: 'displayName', value: 'Topic' } });

        await wait(() => {
            expect(setDisplayNameSpy).toHaveBeenCalledWith('Topic');
        });
    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {
        describe('and the service call is successful', () => {
            it('should redirect to the home page', async () => {
                expect.assertions(1);

                jest.spyOn(TopicsService, 'addParentTopic').mockImplementationOnce(() => Promise.resolve({ response: { status: 200 } }));

                const submitButton = await waitForElement(async () => {
                    return await wrapper.findByText('Submit');
                });
                fireEvent.click(submitButton);

                await wait(() => {
                    expect(history.push).toHaveBeenCalledWith('/', { successMessage: 'The parent topic was created successfully' });
                });
            });
            it('should call the onSubmit action', async () => {
                expect.assertions(1);

                jest.spyOn(TopicsService, 'addParentTopic').mockImplementationOnce(() => Promise.resolve({ response: { status: 200 } }));

                const submitButton = await waitForElement(async () => {
                    return await wrapper.findByText('Submit');
                });
                fireEvent.click(submitButton);

                await wait(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails', () => {
            it('should set the error state', async () => {
                expect.assertions(1);

                jest.spyOn(TopicsService, 'addParentTopic').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));

                const submitButton = await waitForElement(async () => {
                    return await wrapper.findByText('Submit');
                });
                fireEvent.click(submitButton);

                await wait(() => {
                    expect(setMessageSpy).toHaveBeenCalledWith({ description: ADD_PARENT_TOPIC_ERROR_DESCRIPTION, title: GENERAL_ERROR_TITLE });
                });
            });
            it('should call the begin submit action', async () => {
                expect.assertions(1);

                jest.spyOn(TopicsService, 'addParentTopic').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));

                const submitButton = await waitForElement(async () => {
                    return await wrapper.findByText('Submit');
                });
                fireEvent.click(submitButton);

                await wait(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails with a 400', () => {
            it('should set the error state', async () => {
                expect.assertions(1);

                jest.spyOn(TopicsService, 'addParentTopic').mockImplementationOnce(() => Promise.reject({ response: { status: 400 } }));

                const submitButton = await waitForElement(async () => {
                    return await wrapper.findByText('Submit');
                });
                fireEvent.click(submitButton);

                await wait(() => {
                    expect(setMessageSpy).toHaveBeenCalledWith(new ErrorMessage(DUPLICATE_PARENT_TOPIC_DESCRIPTION, VALIDATION_ERROR_TITLE));
                });
            });
        });
    });

    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            useStateSpy.mockImplementationOnce(() => [undefined, setDisplayNameSpy]);

            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect.assertions(1);

            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(addFormErrorSpy).toHaveBeenCalledWith({ key: '', value: 'The Parent Topic is required' });
            });
        });
    });

    describe('with invalid characters', () => {
        beforeEach(async () => {
            useStateSpy.mockImplementationOnce(() => ['Invalid@Topic', setDisplayNameSpy]);

            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect.assertions(1);

            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(addFormErrorSpy).toHaveBeenCalledWith({ key: '', value: 'The Parent Topic contains invalid characters' });
            });
        });
    });

    describe('with valid characters', () => {
        beforeEach(async () => {
            useStateSpy.mockImplementationOnce(() => ['Invalid-Topic', setDisplayNameSpy]);

            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect.assertions(1);

            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(addFormErrorSpy).toBeCalledTimes(0);
            });
        });
    });
});
