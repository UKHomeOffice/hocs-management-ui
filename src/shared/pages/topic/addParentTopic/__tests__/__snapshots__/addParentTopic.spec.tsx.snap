// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addChildTopic component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add parent topic
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        method="POST"
      >
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="topicName"
            id="topicName-label"
          >
            Topic Name
          </label>
          <input
            class="govuk-input"
            id="topicName"
            name="topicName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`when the addParentTopic component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add parent topic
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        method="POST"
      >
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="topicName"
            id="topicName-label"
          >
            Topic Name
          </label>
          <input
            class="govuk-input"
            id="topicName"
            name="topicName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
